### DISCUSSION TITLE
How do I use runtime reflection with functions?

### DISCUSSION THREAD

**Discussion by elliotchance** [vlang/v#11893] [Ideas]
```v
fn add(a int, b int) int {
	return a + b
}

fn call(f any) {
	// How can I introspect f to pass in
	// args and receive the return value?
}

fn main() {
	println(add(5, 3))
	call(add)
}
```

**Comment by elliotchance** [comment]
I wasn't really clear on my question. `call` should not know the definition of the function, it should be able to introspect it can call it at runtime, like in Go:

```go
package main

import (
	"fmt"
	"reflect"
)

func add(a int, b int) int {
	return a + b
}

func call(f interface{}) {
	// Let's assume we have already checked to see the number of args.
	results := reflect.ValueOf(f).Call([]reflect.Value{
		reflect.ValueOf(5),
		reflect.ValueOf(3),
	})
	fmt.Println(results[0].Interface())
}

func main() {
	fmt.Println(add(5, 3))
	call(add)
}
```

**Comment by JalonSolov** [comment]
V supports compile-time reflection.  It does not support runtime reflection.  Runtime reflection usually requires a runtime (which V does not supply), and it can be quite slow (which V does not want to be).

**Comment by elliotchance** [comment]
> V supports compile-time reflection. It does not support runtime reflection. Runtime reflection usually requires a runtime (which V does not supply), and it can be quite slow (which V does not want to be).

Is V ever going to support runtime reflection of anything? What about arrays that may contain mixed fundamental types?

**Comment by JalonSolov** [comment]
It is impossible to say whether it will _ever_ support either of those, but for now, neither one is on any roadmap I've seen.

**Comment by elliotchance** [comment]
So, you can actually do this - at least I have written the working code a while ago for [elliotchance/reflect](https://github.com/elliotchance/reflect) - but not yet merged it.

I am on holiday for the next few days, but I’ll dig it up when I get back and create a PR.

**Reply by mihaigalos** [reply]
Once you have static reflection, you can implement the dynamic one.
Very nice work @elliotchance!

**Reply by ylluminate** [reply]
Anything the community can help you with @elliotchance if you're running into any issues that prevent finishing this up?

**Reply by mihaigalos** [reply]
I've had a look, looks pretty complete with regard to known types at compile time.
Judging by the description below, seems a review is in order here by someone more knowledgeable with the core implementation / values of `vlang`:

> IMPORTANT: This project is more to demonstrate that a lot of the runtime reflection functionality one might need can be done without technically using runtime types. This implementation relies on static code being generated from compile-time reflection for known types going into the appropriate constructors.

Anybody know how difficult it is to construct arbitrary types at runtime if all constituent types are already known (including primitive types)?

**Reply by ylluminate** [reply]
@mihaigalos might be worth taking this to #v-chat or #v-development in https://discord.gg/vlang.

**Reply by mihaigalos** [reply]
I won't take credit for @elliotchance's work. All glory is his! :)
On a side-node, why is a chat necessary in this context? Is the debate not better suited for this thread?

**Reply by elliotchance** [reply]
I just had a look and I must have thrown that sample code away :(

If I remember correctly, you can inspect the argument types (not sure if I did this by stringifying/splitting or if the syntax supported me grabbing these argument types at compile time) then constructing the call by packing arguments into their correct size and offsets in reverse order (arguments are called in reverse internally). Using the function pointer and this crafted binary blob you can do an unsafe call that does the same thing the compiler would construct at compile time.

The biggest danger here I'm sure is how different architectures would pack these values. Especially if they were constructs themselves which can also have optimizations and other memory padding for alignment.

That being said, a better approach would be to have the compiler generate templated proxy functions so that instead of constructing the arguments at runtime you could use map between the signature and a fitting concrete implementation that has already gone through the compiler.

**Reply by mihaigalos** [reply]
> I just had a look and I must have thrown that sample code away :(

That's why tests act as a documentation of how to interact with the code. Please consider extending them instead of providing "sample code".
You can then just reference lines in the test as a form of "look here to see how to".

> The biggest danger here I'm sure is how different architectures would pack these values.

C can be forced to pack struct/arrays in a certain way via `#pragma pack` preprocessor directives. This then guarantees consistent padding. Since V's backend is C anyway, I see no blocker here.

**Reply by JalonSolov** [reply]
C is V's _current_ default backend.  That is due to change once the `-native` backend is completed.  C will still be optional, but `-native` produces much smaller code more quickly, which is what you want for development at least.

V also supports JavaScript (several flavors), and now has support for `wasmer` output.  So yes, this could very well be a blocker unless something is implemented that works with _ALL_ current (and hopefully future) backends.

**Reply by ylluminate** [reply]
(To answer your question of Discord @mihaigalos, I believe that the GH Discussions get's less attention than Discord at this point and I was simply attempting to offer a suggestion to obtain more input since it seems relatively few people respond directly here unless given a link to a Discussion or Issue in Discord. For, for example, my recommendation on this topic would be to paste a link to a Discussion/comment from here into Discord with a request for input such that folks can realize this discussion exists and readily find it.)

**Reply by playXE** [reply]
> C is V's _current_ default backend. That is due to change once the `-native` backend is completed. C will still be optional, but `-native` produces much smaller code more quickly, which is what you want for development at least.
> 
> V also supports JavaScript (several flavors), and now has support for `wasmer` output. So yes, this could very well be a blocker unless something is implemented that works with _ALL_ current (and hopefully future) backends.

V does not have Wasmer output support as Wasmer is just WASM engine that can execute .wasm files. V has built-in support for wasm32-unknown-unknown target that does not have any libc functions provided through wasi or emscripten

**Reply by JalonSolov** [reply]
Fine.  The point is that anything that is implemented needs to support all current and hopefully future backends, rather than specifically targetting C.
